src/api/stats.c

```diff
@@ -160,10 +150,23 @@ int api_stats_top_domains(struct ftl_conn *api)
 		cJSON *json = JSON_NEW_OBJECT();
 		cJSON *top_domains = JSON_NEW_ARRAY();
 		JSON_ADD_ITEM_TO_OBJECT(json, "top_domains", top_domains);
-		free(temparray);
 		JSON_SEND_OBJECT(json);
 	}
 
+	// Lock shared memory
+	lock_shm();
+
+	// Allocate memory
+	int count = 10;
+	bool audit = false;
+	const int domains = counters->domains;
+	int *temparray = calloc(2*domains, sizeof(int*));
+	if(temparray == NULL)
+	{
+		log_err("Memory allocation failed in %s()", __FUNCTION__);
+		return 0;
+	}
```

Comment:
This code has been moved down to after the `if` so we don't even have to allocate the memory when exiting anyway due to privacy settings. I also moved the shared memory locking up so we can be sure `counter->domains` does not change while we are doing this here (race-collision prevention).